shell/renderer/api/electron_api_web_frame.cc

```diff
@@ -130,10 +134,54 @@ class ScriptExecutionCallback : public blink::WebScriptExecutionCallback {
     v8::Isolate* isolate = v8::Isolate::GetCurrent();
     if (!result.empty()) {
       if (!result[0].IsEmpty()) {
-        // Right now only single results per frame is supported.
-        if (!callback_.is_null())
-          std::move(callback_).Run(result[0], v8::Undefined(isolate));
-        promise_.Resolve(result[0]);
+        // Either world safe results are disabled or the result was created in
+        // the same world as the caller
+        if (!world_safe_result_ ||
+            (result[0]->IsObject() &&
+             promise_.GetContext() ==
+                 result[0].As<v8::Object>()->CreationContext())) {
+          // Right now only single results per frame is supported.
+          if (!callback_.is_null())
+            std::move(callback_).Run(result[0], v8::Undefined(isolate));
+          promise_.Resolve(result[0]);
+        } else {
+          // Serializable objects
+          blink::CloneableMessage ret;
+          bool success;
+          std::string error_message;
+          {
+            v8::TryCatch try_catch(isolate);
+            success = gin::ConvertFromV8(isolate, result[0], &ret);
+            if (try_catch.HasCaught()) {
+              auto message = try_catch.Message();
+
+              if (message.IsEmpty() ||
+                  !gin::ConvertFromV8(isolate, message->Get(),
+                                      &error_message)) {
+                error_message =
+                    "An unknown exception occurred while getting the result of "
+                    "the script";
+              }
+            }
+          }
+          if (!success) {
+            // Failed convert so we send undefined everywhere
+            if (!callback_.is_null())
+              std::move(callback_).Run(
+                  v8::Undefined(isolate),
+                  v8::Exception::Error(
+                      v8::String::NewFromUtf8(isolate, error_message.c_str())
+                          .ToLocalChecked()));
+            promise_.RejectWithErrorMessage(error_message);
+          } else {
+            v8::Local<v8::Context> context = promise_.GetContext();
+            v8::Context::Scope context_scope(context);
+            v8::Local<v8::Value> cloned_value = gin::ConvertToV8(isolate, ret);
+            if (!callback_.is_null())
+              std::move(callback_).Run(cloned_value, v8::Undefined(isolate));
+            promise_.Resolve(cloned_value);
+          }
+        }
```

Comment:
getting highly indented here. Can we break some of this functionality out into a helper function?